### PR TITLE
Sync refactor backend [WIP]

### PR DESCRIPTION
**Pull request by camsaul**
Backend changes for sync refactoring. From #5363

### CODE REVIEW

File: test/metabase/models/field_test.clj [hunk 1]
```diff
@@ -4,82 +4,3 @@
...
-
-(expect nil            (infer-field-special-type nil       nil))
-(expect nil            (infer-field-special-type "id"      nil))
-(expect nil            (infer-field-special-type nil       :type/Integer))
-(expect :type/PK       (infer-field-special-type "id"      :type/Integer))
-;; other pattern matches based on type/regex (remember, base_type matters in matching!)
-(expect :type/Category (infer-field-special-type "rating"  :type/Integer))
-(expect nil            (infer-field-special-type "rating"  :type/Boolean))
-(expect :type/Country  (infer-field-special-type "country" :type/Text))
-(expect nil            (infer-field-special-type "country" :type/Integer))
```
Comment (camsaul) [inline test/metabase/models/field_test.clj]: things you should never do:

delete all the code from a namespace and then leave the empty namespace there for no reason

File: test/metabase/api/table_test.clj [hunk 2]
```diff
@@ -455,8 +484,11 @@
                                                               :updated_at   $
                                                               :id           $
                                                               :raw_table_id $
-                                                              :created_at   $}))}))}])
-  ((user->client :rasta) :get 200 (format "table/%d/fks" (id :users))))
+                                                              :created_at   $}))}))}]))
+(do (fingerprint/cache-field-values-for-table! (Table (id :users)))
```
Comment (camsaul) [inline test/metabase/api/table_test.clj]: other bad ideas:

treating code indentation as optional and just starting from column 1 when you feel like it

File: resources/migrations/000_migrations.yaml [hunk 3]
```diff
@@ -3670,3 +3670,189 @@ databaseChangeLog:
                   name: result_metadata
                   type: text
                   remarks: 'Serialized JSON containing metadata about the result columns from running the query.'
+  - changeSet:
+      id: 58
+      author: arthur
+      changes:
+        - createTable:
+            tableName: field_fingerprint
+            remarks: 'tables are sampled periodically to capture a sample of the values in them. These snapshots are then used to make these statistics which will later be used to guess properties of that table'
```
Comment (camsaul) [inline resources/migrations/000_migrations.yaml]: another bad idea:

copying and pasting docstrings from other pieces of code and not updating them, which leaves you with documentation that doesn't make sense

File: src/metabase/sfc/classify.clj [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+(defn test:url-special-type
+  "If FIELD is texual, doesn't have a `special-type`, and its non-nil values are primarily URLs, mark it as `special-type` `:type/URL`."
+  [fingerprint field-stats]
+  (if-not (and (not (:special-type field-stats))
+               (isa? (:base_type fingerprint) :type/Text))
+    ;; this field isn't suited for this test
+    field-stats
+    ;; test for url values
+    (let [percent-urls (:field-percent-urls fingerprint)]
```
Comment (camsaul) [inline src/metabase/sfc/classify.clj]: @salsakran do you have a sense if any of this code ever worked? I keep stumbling upon new things like this when I read the code. Here we're trying to look up the key `:field-percent-urls` in the fingerprint. However, the fingerprint uses the key `:field_percent_urls`

File: src/metabase/sfc/analyze.clj [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+(defn field-fingerprint [driver table field]
+  (let [values      (->> (driver/field-values-lazy-seq driver field)
+                         (take driver/max-sync-lazy-seq-results))
+        field-id    (:id field)
+        fingerprint {:base_type           (:base_type field)
+                     :is_pk               (isa? (:special_type field) :type/PK)
+                     :is_fk               (isa? (:special_type field) :type/FK)
+                     :cardinality         (count (distinct values))
+                     :field_percent_urls  (percent-valid-urls values)
```
Comment (camsaul) [inline src/metabase/sfc/analyze.clj]: @salsakran `:field_percent_urls` is set here

File: src/metabase/sfc/analyze.clj [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+   :table_id id
+   :name     name})
+
+(defn- save-table-fingerprint!
+  "store the table fingerprint for a table
+   field values are stored separately"
+  [fingerprint]
+  (or (db/update! TableFingerprint {:where [:= :table_id (:table_id fingerprint)]
+                                    :set fingerprint})
+      (db/insert! TableFingerprint fingerprint)))
```
Comment (camsaul) [inline src/metabase/sfc/analyze.clj]: as mentioned in the other PR this is a potential race condition. Should fix this and the similar code above

File: test/metabase/sfc/sync_test.clj [hunk 7]
```diff
@@ -292,43 +306,9 @@
...
-     ;; 3. Now re-sync the table and make sure the value is back
-     (do (sync-table! @venues-table)
-         (get-field-values))]))
-
-
-;;; -------------------- Make sure that if a Field's cardinality passes `metabase.sync-database.analyze/low-cardinality-threshold` (currently 300) (#3215) --------------------
+;;; -------------------- Make sure that if a Field's cardinality passes `metabase.sfc.classify/low-cardinality-threshold` (currently 300) (#3215) --------------------
 (defn- insert-range-sql [rang]
-  (str "INSERT INTO blueberries_consumed (num) VALUES "
+  (str "INSERT INTO blueberries_consumed_rating (num) VALUES "
```
Comment (camsaul) [inline test/metabase/sfc/sync_test.clj]: TODO - maybe change this test data name? The new name makes zero sense

File: src/metabase/task/sync.clj [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        (log/error (format "Error fetching field values for database %d: (%s)" db-id (:name database)) e)))))
+
+(jobs/defjob SyncDatabase [job-context]
+  (let [db-id    (get (qc/from-job-data job-context) "db-id")
+        database (Database db-id)]
+    (try
+      (log/debug (u/format-color 'green "running scheduled caching of field values for database-id: %s: %s" db-id database))
+      (fingerprint/cache-database-field-values! database :full-sync? true)
+      (catch Throwable e
+        (log/error (format "Error fetching field values for database %d: (%s)" db-id (:name database)) e)))))
```
Comment (camsaul) [inline src/metabase/task/sync.clj]: @salsakran you wouldn't happen to know what the purpose of this `SyncDatabase` task is, would you? It does basically the same exact thing as the `CacheFieldValuesForDatabase` task, and doesn't actually do any syncing.

Comment (salsakran) [inline src/metabase/task/sync.clj]: negatory. It *should* just do the sync process.

File: src/metabase/task/sync.clj [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+(defonce ^:private classify-databases-job (atom nil))
+
+(defn task-init
+  "classify called during startup; start the job for classify databases."
+  []
+  ;; build one job and one trigger for each database.
+  (let [triggers (doseq [database (db/select Database, :is_sample false)]
+                   (schedule-db-sync-actions! database))] ;; we're building a sequence of these jobs so they can be stopped later
+    (reset! classify-databases-job triggers)))
```
Comment (camsaul) [inline src/metabase/task/sync.clj]: @salsakran I found some incomplete code here. Here when we schedule all the DB sync tasks we store the names of all the jobs that were created. The comment suggests

> we're building a sequence of these jobs so they can be stopped later

However, there's no code anywhere that reads that list of job names or stops them at any point. 

What I'm wondering is, was this feature something that needs to ship as part of this PR? Or is it something that I can take out the dead code for (we can always add it back later if we want to add that feature)

Comment (salsakran) [inline src/metabase/task/sync.clj]: This PR's priorities:
1. allow for separate schedules of fields getting their values cached vs getting analyzed/fingerprinted
2. Allow for an admin facing UX to set the field value caching schedule + allow manual re-caches of fields + databases
3. Allow value caching to happen via samples. Bonus points for allowing a "do a tablescan" override on this.

**Comment by camsaul** [comment]
good news and bad. I got tests to pass. Bad news is there is some code that is clearly wrong that I didn't fix when it passed. That means we are definitely missing a LOT of tests here. Yay

**Comment by camsaul** [comment]
Superseded by #5510

**Pull request closed without merging**